pkg/distributor/otlpappender/combined_appender.go

```diff
@@ -0,0 +1,203 @@
+	refs          map[uint64]labelsIdx
+	collisionRefs map[uint64][]labelsIdx
+
+	// metricFamilies is used to store metadata for each metric family.
+	// This is needed to not send metadata duplicates all the time.
+	// We could get rid of this if we switched to RW2 all the way through.
+	metricFamilies map[string]metadata.Metadata
+}
+
+func NewCombinedAppender() *CombinedAppender {
+	return &CombinedAppender{
+		ValidIntervalCreatedTimestampZeroIngestion: defaultIntervalForStartTimestamps,
+		series:         mimirpb.PreallocTimeseriesSliceFromPool(),
+		refs:           make(map[uint64]labelsIdx),
+		collisionRefs:  make(map[uint64][]labelsIdx),
+		metricFamilies: make(map[string]metadata.Metadata),
+	}
+}
+
+// GetResult returns the created timeseries and metadata.
+func (c *CombinedAppender) GetResult() ([]mimirpb.PreallocTimeseries, []*mimirpb.MetricMetadata) {
+	return c.series, c.metadata
+}
+
+func (c *CombinedAppender) AppendSample(ls labels.Labels, meta otlpappender.Metadata, ct, t int64, v float64, es []exemplar.Exemplar) error {
+	return c.appendFloatOrHistogram(ls, meta, ct, t, v, nil, es)
+}
+
+func (c *CombinedAppender) AppendHistogram(ls labels.Labels, meta otlpappender.Metadata, ct, t int64, h *histogram.Histogram, es []exemplar.Exemplar) error {
+	return c.appendFloatOrHistogram(ls, meta, ct, t, 0, h, es)
+}
+
+func (c *CombinedAppender) appendFloatOrHistogram(ls labels.Labels, meta otlpappender.Metadata, ct, t int64, v float64, h *histogram.Histogram, es []exemplar.Exemplar) error {
+	ct = c.recalcCreatedTimestamp(t, ct)
+
+	hash, idx, collisionIdx, seenSeries := c.processLabelsAndMetadata(ls)
+
+	if !seenSeries || c.ctRequiresNewSeries(idx.idx, ct) {
+		c.createNewSeries(&idx, collisionIdx, hash, ls, ct)
+	}
+
+	if h != nil {
+		c.series[idx.idx].Histograms = append(c.series[idx.idx].Histograms, mimirpb.FromHistogramToHistogramProto(t, h))
+	} else {
+		c.series[idx.idx].Samples = append(c.series[idx.idx].Samples, mimirpb.Sample{TimestampMs: t, Value: v})
+	}
+	c.appendExemplars(idx.idx, es)
+	c.appendMetadata(meta.MetricFamilyName, meta.Metadata)
+
+	return nil
+}
+
+func (c *CombinedAppender) recalcCreatedTimestamp(t, ct int64) int64 {
+	if !c.EnableCreatedTimestampZeroIngestion || ct < 0 || ct > t || (c.ValidIntervalCreatedTimestampZeroIngestion > 0 && t-ct > c.ValidIntervalCreatedTimestampZeroIngestion) {
+		return 0
+	}
+
+	return ct
+}
+
+// ctRequiresNewSeries checks if the created timestamp is meaningful and different
+// from the one already stored in the series at the given index.
+func (c *CombinedAppender) ctRequiresNewSeries(seriesIdx int, ct int64) bool {
+	return ct > 0 && c.series[seriesIdx].CreatedTimestamp != ct
+}
+
+// processLabelsAndMetadata figures out if we have already seen this
+// exact label set and whether we need to update the metadata.
+// The returned collisionIdx is 0 is there's no hash collision and
+// index+1 into the collisions otherwise.
```

Comment:
To avoid confusion/programmer mistakes, can we use normal zero based indexing for collisions?
```suggestion
// The returned collisionIdx is -1 if there's no hash collision and
// an index into the collisions otherwise.
```

Reply:
done in b825954cfad6a6d9daae30337a5148d9c9d9239d